src/util/irep.h

```diff
@@ -82,8 +82,77 @@ inline const std::string &name2string(const irep_namet &n)
 
-/*! \brief Base class for tree-like data structures with sharing
-*/
+/// \brief Base class for tree-like data structures with sharing
+///
+/// There are a large number of kinds of tree structured or tree-like data in
+/// CPROVER. `irept` provides a single, unified representation for all of
+/// these, allowing structure sharing and reference counting of data. As
+/// such `irept` is the basic unit of data in CPROVER.  Each `irept`
+/// contains[^1] a basic unit of data (of type `dt`) which contains four
+/// things:
+///
+/// * `data`:   A string[^2], which is returned when the `id()` function is
+///   used.
+///
+/// * `named_sub`:   A map from `irep_namet` (a string) to `irept`. This
+///   is used for named children, i.e.  subexpressions, parameters, etc.
+///
+/// * `comments`:   Another map from `irep_namet` to `irept` which is used
+///   for annotations and other ‘non-semantic’ information
+///
+/// * `sub`:   A vector of `irept` which is used to store ordered but
+///   unnamed children.
+///
+/// The `irept::pretty` function outputs the contents of an `irept` directly
+/// and can be used to understand and debug problems with `irept`s.
+///
+/// On their own `irept`s do not “mean” anything; they are effectively
+/// generic tree nodes. Their interpretation depends on the contents of
+/// result of the `id` function (the `data`) field. `util/irep_ids.txt`
+/// contains the complete list of `id` values. During the build process it
+/// is used to generate `util/irep_ids.h` which gives constants for each id
+/// (named `ID_`). These can then be used to identify what kind of data
+/// `irept` stores and thus what can be done with it.
+///
+/// To simplify this process, there are a variety of classes that inherit
+/// from `irept`, roughly corresponding to the ids listed (i.e.  `ID_or`
+/// (the string `"or”`) corresponds to the class `or_exprt`). These give
+/// semantically relevant accessor functions for the data; effectively
+/// different APIs for the same underlying data structure. None of these
+/// classes add fields (only methods) and so static casting can be used. The
+/// inheritance graph of the subclasses of `irept` is a useful starting
+/// point for working out how to manipulate data.
+///
+/// There are three main groups of classes (or APIs); those derived from
+/// `typet`, `codet` and `exprt` respectively. Although all of these inherit
+/// from `irept`, these are the most abstract level that code should handle
+/// data. If code is manipulating plain `irept`s then something is wrong
+/// with the architecture of the code.
+///
+/// Many of the key descendant of `exprt` are declared in `std_expr.h`. All
+/// expressions have a named subfield / annotation which gives the type of
+/// the expression (slightly simplified from C/C++ as `unsignedbv_typet`,
+/// `signedbv_typet`, `floatbv_typet`, etc.). All type conversions are
+/// explicit with an expression with `id() == ID_typecast` and an ‘interface
```

Comment:
"interface class" is not used anywhere else, just say "i.e. a `typecast_exprt`" instead